### PR TITLE
Feature request: use cotire

### PR DESCRIPTION
**Pull request by nmellado**
* **Please check if the PR fulfills these requirements**
- [x] Is the Pull-Request done against the right branch:
  - `master-v1`: for changes related to the Radium Stable Release V1.
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Update buildchain to use cotire (precompiled headers) to speed up the compilation.
Feature request #172 

* **What is the current behavior?** (You can also link to an open issue here)
There is no precompiled header system.

* **What is the new behavior (if this is a feature change)?**
Cotire is used to precompile headers on `Ra::Core`, `Ra::IO` (except for MacOSx, which is buggy) and the apps packaged with Radium.

* **Does this PR introduce a breaking change?** 
 No

* **Other information**
Commits should be stashed prior rebase.

### CODE REVIEW

File: src/IO/AssimpLoader/AssimpGeometryDataLoader.hpp [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 
-#include <set>
-
 #include <Core/File/DataLoader.hpp>
 #include <Core/Math/LinearAlgebra.hpp>
 #include <IO/RaIO.hpp>
 
+#include <memory>
+#include <set>
+
```
Comment (MathiasPaulin) [inline src/IO/AssimpLoader/AssimpGeometryDataLoader.hpp]: why are these needed here ?

Comment (nmellado) [reply src/IO/AssimpLoader/AssimpGeometryDataLoader.hpp]: `set` was already here, and `memory` is here for `unique_ptr`.

**Comment by nmellado** [comment]
Cotire is enabled on MacOS, except for IO.
Which problem do you see with the way cotire is used ? (I followed the documentation)

File: src/IO/AssimpLoader/AssimpGeometryDataLoader.hpp (hunk 1 above)
Comment (MathiasPaulin) [reply src/IO/AssimpLoader/AssimpGeometryDataLoader.hpp]: OK. But I've strange compilation warning (and sometime errors). I investigate ...

**Comment by nmellado** [comment]
I will try to see how to add GCC on MacOS on Travis, this could help to detect such problems.

**Comment by MathiasPaulin** [comment]
I made a make clean before branch switching then clear the cmake cache and do a cmake .....
I'll try to remove the complete Bundle directory ...

**Comment by MathiasPaulin** [comment]
After removing the whole bundle and all the cmake generated files and directory (rm -rf cmake-build-*), I succes in compiling with AppleClang 9.0.0.9000039. 
Note that I reactivated cotire on IO and the compilation successes.
Here are, nevertheless,  the list of cotire-based warnings I got 
- Cmake configuration : 
```
Problems were encountered while collecting compiler information:
	<built-in>:1:10: fatal error: '/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/Applications/HelloRadium/cotire/helloRadium_CXX_prefix.hxx' file not found
	<built-in>:1:10: fatal error: '/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/Applications/MainApplication/cotire/main-app_CXX_prefix.hxx' file not found
	<built-in>:1:10: fatal error: '/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/cotire/radiumCore_CXX_prefix.hxx' file not found
	<built-in>:1:10: fatal error: '/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/cotire/radiumIO_CXX_prefix.hxx' file not found
```
- when compiling : 
```
-- Result 1 scanning includes of /Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/cotire/radiumIO_CXX_unity.cxx.
-- 5 line(s) skipped, see /Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/cotire/radiumIO_CXX_prefix.cxx.log
```
with the content of the mentioned file 
```
In file included from /Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/cotire/radiumIO_CXX_unity.cxx:4:
/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/src/IO/CameraLoader/CameraLoader.cpp:1:10: fatal error: 'IO/CameraLoader/CameraLoader.hpp' file not found
#include <IO/CameraLoader/CameraLoader.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
But this error is not fatal as the compilation success and all the applications are OK.

**Comment by MathiasPaulin** [comment]
Well, sometime (I don't understand why nor under which circumstances), I got the following error 
```
/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/Version.cpp:15:26: error: redefinition of 'const char* Ra::Version::compileDate'
 const char* Ra::Version::compileDate  = __DATE__;
                          ^~~~~~~~~~~
/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/Version.cpp:2:13: note: 'const char* Ra::Version::compileDate' previously defined here
 const char* Ra::Version::compileDate  = __DATE__;
             ^~
/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/Version.cpp:16:26: error: redefinition of 'const char* Ra::Version::compileTime'
 const char* Ra::Version::compileTime  = __TIME__;
                          ^~~~~~~~~~~
/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/Version.cpp:3:13: note: 'const char* Ra::Version::compileTime' previously defined here
 const char* Ra::Version::compileTime  = __TIME__;
             ^~
/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/Version.cpp:17:26: error: redefinition of 'const char* Ra::Version::gitChangeSet'
 const char* Ra::Version::gitChangeSet = "a1186d57c6f0b83eae7f2195c5f19591ebc830d3";
                          ^~~~~~~~~~~~
/Users/mathias/Professionnel/SVN-IRIT-paulin/paulin/Developpement/Radium/Radium-Engine/cmake-build-release/src/Version.cpp:7:13: note: 'const char* Ra::Version::gitChangeSet' previously defined here
 const char* Ra::Version::gitChangeSet = "a1186d57c6f0b83eae7f2195c5f19591ebc830d3";
```
It seems that there are concurrent writing on this generated file, I also saw scrambled lines on another compilation ...

**Comment by nmellado** [comment]
@MathiasPaulin , regarding your comment https://github.com/STORM-IRIT/Radium-Engine/pull/367#issuecomment-425119645
I've tried to protect against this problem by doing this:
```cmake
set_target_properties(radiumVersionFile PROPERTIES COTIRE_ENABLE_PRECOMPILED_HEADER FALSE)
```

This target is defined as a dependency for radium Core library:
```cmake
add_dependencies( ${core_target} ... radiumVersionFile)
cotire(${core_target})
```
For me it fixed the problem.
To be honest I don't exactly know why it is still wrong on your side. I cannot find any other thing in the documentation. 
We might need to do a minimal example and open an issue on cotire repository.

**Comment by nmellado** [comment]
@MathiasPaulin @dlyr I need your help here to configure g++8 on macos for the CI tests.

**Comment by MathiasPaulin** [comment]
for the problem of generating Version.cpp twice at the same time (and then, get an erroneous file), I found this https://www.mail-archive.com/cmake@cmake.org/msg32782.html
There is no definitive conclusion in this discussion but I fix the problem by removing the custom_target radiumVersionFile and the dependency of radium_core on this target. The generated file only appears now on 
```
add_library(
    ${core_target} SHARED
    ${core_sources}
    ${core_headers}
    ${core_inlines}
    ${RADIUM_VERSION_FILE} #This file is build-dependant and generated in the Bundle
)
```
Seems to work well for me, I'll push the commit as soon as I've verified that this does not break something else.

File: .travis.yml [hunk 2]
```diff
@@ -89,15 +69,26 @@ matrix:
...
       compiler: clang
 
+#    # mac gcc Release
+#    - env:
+#        - BUILD_TYPE=Release
+#        - PREFIX_PATH="/usr/local/opt/qt/lib/cmake"
+#        - BUILD_DOC=Off
+#	# see https://github.com/Homebrew/brew/issues/1742
+#        - MATRIX_EVAL="rm /usr/local/include/c++ && brew install gcc && CC=gcc-8 && CXX=g++-8"
+#      os: osx
```
Comment (dlyr) [inline .travis.yml]: Why is this commented ? if not working, might just remove theses lines ?

Comment (nmellado) [reply .travis.yml]: Fixed by 8eb878c
